### ISSUE TITLE
Template defaults

### ISSUE THREAD

**Issue by alexec** [argoproj/argo-workflows#5148]
# Summary

Users sometimes want to set default values for templates. We do this piecemeal today, e.g. by adding a new field under `spec`, but we could do this in a general way, e.g.

```yaml
spec:
  templateDefaults:
    memoization:
      cache: "cache-name-v1"
```

Note that template defaults should be parameterization (e.g. using `fasttemplate`).

# Use Cases

To reduce duplication.

See #5113
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by terrytangyuan** [comment]
Looks great and this solution is more generalized. What's the relationship of this with those workflow-level fields like `WorkflowSpec.RetryStrategy`? Perhaps the template defaults will get overridden by workflow-level fields?

**Comment by simster7** [comment]
We're thinking of adding some sort of `strategy:` field. `Override` could be a setting, as well as `Merge`. In the case of `memoization`, for example, `Merge` would be more applicable

**Comment by sarabala1979** [comment]
resolved by #5410
